[spirv] Add initial I64 emulation pass

kuhar opened the pull request:
The pass is conditionally enabled when the SPIR-V target does not declare I64 capabilities. The `demote-i64-to-i32' needs to be disabled for Wide Integer Emulation (WIE) to have any i64 ops to emulate. These default will be changed once the WIE implementation is robust enough.

Locally, I tested this by changing the `demote-i64-to-i32` default to false and all tests passed.

Issue: #8702

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/SPIRVEmulateI64.cpp:
```diff
@@ -0,0 +1,200 @@
+static bool supportsI64(ModuleOp op) {
+  spirv::TargetEnvAttr targetEnv = getSPIRVTargetEnvAttr(op);
+  assert(targetEnv && "Not a valid spirv module");
+  return llvm::is_contained(targetEnv.getCapabilities(),
```
We should create a `spirv::TargetEnv` out of the `spirv::TargetEnvAttr` to use the `allows` API there. The difference is, `spirv::TargetEnvAttr` only gives us the exact same set of capabilities specified by the environment, but certain capabilities are _implied_ (e.g., `Matrix` is always implied with `Shader`). `spirv::TargetEnv` handles the implication relationship under the hood.

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/SPIRVEmulateI64.cpp:
```diff
@@ -0,0 +1,200 @@
+//===----------------------------------------------------------------------===//
+// Conversion patterns
+//===----------------------------------------------------------------------===//
+struct ConvertHalInterfaceBindingSubspan final
```
We need to add IR lit tests for the new patterns added here. It allows to guard against breakages and identify issues quicker. (End-to-end correctness runtime tests would involve more steps than working on IR lit tests.)

antiagainst reviewed (changes requested):
Thanks! We'd need to have lit tests for the patterns. Another thing is that cherry-picking upstream commits should happen in its own pull request (so it's easier for integration folks to know and easier to revert if there are issues)

antiagainst commented:
>  Another thing is that cherry-picking upstream commits should happen in its own pull request (so it's easier for integration folks to know and easier to revert if there are issues)

Not hard requirement though. And there are cases we need to fix iree to make it compile/pass. So use your judgement. :)

kuhar commented:
Thanks for the feedback, @antiagainst! Addressed comments and added new LIT test. PTAL.

> > Another thing is that cherry-picking upstream commits should happen in its own pull request (so it's easier for integration folks to know and easier to revert if there are issues)
> 
> Not hard requirement though. And there are cases we need to fix iree to make it compile/pass. So use your judgement. :)

I was hoping these cherry-picked commits will propagate to the submodule by the time this PR is ready to land. If the next integrate does not pick them, I'll follow your suggestion.

kuhar commented:
Ping

antiagainst commented on compiler/src/iree/compiler/Codegen/SPIRV/SPIRVEmulateI64.cpp:
```diff
@@ -0,0 +1,207 @@
+
+//===----------------------------------------------------------------------===//
+//
+// This file implements a pass to emualte 64-bit integer operations with 32-bit
```
typo: emualte

antiagainst commented on tests/e2e/vulkan_specific/BUILD:
```diff
@@ -44,3 +44,23 @@ iree_check_single_backend_test_suite(
+    tags = [
+        "manual",
+        "notap",
+        "vulkan_uses_vk_khr_shader_int64",
```
Just `vulkan_uses_shader_int64`? This is not behind an `VK_KHR` extension.
